### PR TITLE
LaTeX Style Guidelines

### PR DESCRIPTION
**Pull request by mdclyburn**
Initial draft for LaTeX style guidelines.

Feedback appreciated.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+## Submission
+
+Before submitting a section for inclusion in the Hack Pack, ensure that both the Hack Pack and Hack Pack++ build (using only `make`) on the McAdams lab machines _without stopping_.
+
+## Still Not Sure?
+
+We strive to make our content as clear and concise as possible, but it is always possible that we have left some content off.
+If you have any question about any part of your LaTeX, it is always a good idea to consult entries that have already made it into the Hack Pack.
+In particular, the section on INSERT SECTION HERE is an example of an excellent addition to the Hack Pack.
```
Comment (robertu94) [inline CONTRIBUTING.md]: I have been telling people `set` is the section to look for, but we should double check that all of these rules are followed there before merging.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+Failing that, feel free to email any of the core Hack Pack team.
```
Comment (robertu94) [inline CONTRIBUTING.md]: I would refer to the group as the `@clemsonacm/hackpack-developers` group

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+### Citations
+
+Information pulled from other sources needs to be properly cited.
+Any information included in submissions that is not properly cited will cause the entire section to be rejected from consideration for inclusion.
+References should use BibTeX where applicable.
+
+## Format
+
+### General Rules to Type By
```
Comment (robertu94) [inline CONTRIBUTING.md:141]: We should include some reasons for these rules. for example:
- Only one line per sentence.  This makes it easier to review changes to documentation.
- Use tabs for indentation.  This allows editors to choose their preferred tab width.
- Do not use LaTeX comments.  Anything important enough to put in a comment should be put in the contents.  Exceptions would include the `style.tex` file.

I also think that we should address when indentation should be used.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+If adding a visual to a section would greatly aid in the understanding of the topic, doing so is highly encouraged.
+Submissions should preferably be in the Encapsulated PostScript format.
+This provides very clear visuals, even when scaled up or down.
+Do _not_ simply pull images from the Internet.
+Doing so may violate the Hack Pack licensing terms.
+Any submission found to be doing so will be rejected until the appropriate changes are made.
+
+### Indexing
+
```
Comment (robertu94) [inline CONTRIBUTING.md:121]: We should also mention that if a term is already in the index because it is marked in another section, it should also be marked for indexing in yours.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+##### Lessons Learned
+
+An unordered list of at least one non-trivial observation you had while writing the solution to your sample problem.
+
+***
+
+### Figures and Images
+
+If adding a visual to a section would greatly aid in the understanding of the topic, doing so is highly encouraged.
+Submissions should preferably be in the Encapsulated PostScript format.
```
Comment (robertu94) [inline CONTRIBUTING.md:109]: Perhaps a comment on useful tools for editing EPS files

File: CONTRIBUTING.md (hunk 2 above)
Comment (mdclyburn) [inline CONTRIBUTING.md]: Would contributors know who to email from that piece of information?

**Comment by robertu94** [comment]
We also need to address minification guidelines in this document

**Comment by mdclyburn** [comment]
Can you list what else there needs to be besides the condensing rules that are already there?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+A complete, condensed section of the Hack Pack should include:
+
+- a brief description of the problem (possibly gleaned from the full description)
+- reference code (if the section is for a data structure)
+- sample solutions for each included problem
+
+### How To Condense Your Section
+
+We have defined special keywords (from a familiar language) to separate information that should be displayed in each section: `#define hackpack`, `#define hackpackpp`, and `#endif`.
```
Comment (robertu94) [inline CONTRIBUTING.md]: It's actually `#ifdef hackpack` and `#ifdef hackpackpp`

**Comment by robertu94** [comment]
The part about attempting to leave in index entries if possible.

Also the minified version should include the list of applications.

**Comment by robertu94** [comment]
I am quite happy with the style guide that we have come up with here.  I propose that we go ahead and merge this.

**Comment by ProtractorNinja** [comment]
- Can you make a note that "Encapsulated PostScript" files are just `.eps` files?
- The bulleted points aren't composed consistently, although I get the impression that they're intentionally done differently in certain cases. The place that stands out the most is the list under "Things to Keep In Mind", which may as well be replaced with a clone of the General Rules part below. I could be wrong.
- I'm not sure this is the place to discuss this, but why are we forcing sample code to live in an example contest problem? This structure fits as a learning resource, but honestly, our goal right now should be to optimize for a contest environment. Wouldn't it be more useful to have a bare-bones DS/algorithm implementation for quick reference?

  All the sample problem IO/format/etc. junk takes up a **lot** of space, complicates things when you're trying to find an algorithm during a contest -- it's tough enough thinking about the problem you're currently trying to solve -- and, generally speaking, requiring a sample problem makes contribution more challenging. They just aren't very useful for a contest environment.

  Hence, I think that Algorithm and Data Structure sections should only require Description and Reference sections ("Applications", if actually relevant, should be a sub-section under Description), with Sample Problems as an **optional** extra for the HackPack++ except in cases of extreme relevance.

**Comment by mdclyburn** [comment]
- Heh heh. My over-formality. I will add the `.eps` note.
- I'll look into this and see what I find. 
- Oh, boy. Another thing I'd be willing to have an extended conversation about. Perhaps we should open another issue? If you don't, I will. 

Thanks a bundle, @ProtractorNinja

**Comment by ProtractorNinja** [comment]
I'll make one. Hang on.

**Comment by robertu94** [comment]
I am fine with making Applications a subsection of Description; however, I still believe that it is needed:

-  It provides a short list of how this is used.  Sure you and I both know that Dijkstra solves shortest paths, but did you know that flow helps you find a minimum path decomposition in a DAG?  Furthermore, with data structures, this is a more useful question to answer.
-  It provides a handy place to include index entries that aren't hidden in a paragraph swamp.

In short, Applications should answer the question, "Why do I care?"

**Comment by ProtractorNinja** [comment]
I went back to review all of the "Applications" sections, and realized that they're sometimes useful, sometimes not -- for instance, the Graph and Heap chapters list applications in a different sense than what you described. Instead of noting, er, actual applications, they vaguely reference algorithms that use the structure. That's not helpful at all, especially if the referenced material isn't actually in the hack pack.

As neat as it is to know that the Bellman Ford algorithm involves graphs with some negative edges, that information is insultingly useless in a contest environment if I still don't know how to use the Bellman Ford algorithm.

"Applications" should be *useful* somehow. If you can't actually apply them, then why are they there?

**Comment by mdclyburn** [comment]
OK. I think I see what you mean @ProtractorNinja. Perhaps the 'Quick Start' section should be moved to the README.md and the CONTRIBUTING.md should remain a comprehensive guide to submitting content?

**Comment by ProtractorNinja** [comment]
I don't think so, actually -- the Readme should link to CONTRIBUTING, but CONTRIBUTING should be a one-stop shop for all things contribution. A quick reference on the top is handy.

I think my confusion was because 'Things to Keep in Mind' section further reduces something from later on that's already pretty quick to read. I'm not sure how to balance them properly.

**Comment by mdclyburn** [comment]
OK. I understand.

**Comment by ProtractorNinja** [comment]
The LaTeX contribution document still suffers due to the as-yet-unaddressed large-scale issues that I wrote about in #63. However, publishing a LaTeX standards guide doesn't need to be held up by those issues--here's some feedback to help get the guide out there:
- The quick start is indeed quick, but somehow feels detached. Perhaps this feeling could be alleviated by rewriting it to be a blend of a quick start and a friendly introduction to the document. As is, the whole thing could benefit from sounding more relaxed.
- I'm still bothered by how there's a mix of bullet grammar and punctuation in contrast to the rest of the document. Can you change the bullets (and context, if necessary) to be properly punctuated?
- The reference section description implies that algorithm sample code is not necessary, but including minimal code would be helpful. This may be a point for #63.

**Comment by mdclyburn** [comment]
Oh, feedback! 
- I'll look into grooming it into something with a silky-smooth, alluring texture. 
- Will do. 
- Yeah... about that... (comment below) 

Sections of this guide almost definitely depend on what makes for a valid contribution. Until #63 is resolved, parts of this document are in a kind of limbo. We need to know exactly what is going to be contributed before we can decisively know how we want it contributed.

**Comment by mdclyburn** [comment]
All right, @ProtractorNinja. So, I've made some changes based on what you have commented on. Of course, you're always welcome to make changes to the document yourself. If the new quick start section is distasteful in any way, please feel free to write it yourself.

This guide _will_ change once we get through #63.

**Comment by ProtractorNinja** [comment]
Looks a lot better. Thanks!
